Basic profile query commands: list and show

mropert opened the pull request:
I've added a simple conan command to query available profiles:
- list: show the list of available profiles
- show <profile>: displays the profile

I'm not too fond of the [_root] hack in Scopes but I'm not certain I master Conan's code enough to propose a better alternative.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/conan-io/conan?pullRequest=734) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/conan-io/conan?pullRequest=734) before we can accept your contribution.<br/>

memsharded commented on conans/client/client_cache.py:
```diff
@@ -164,4 +164,6 @@ def load_profile(self, name):
 
     def current_profiles(self):
+        if not os.path.isdir(self.profiles_path):
+            return []
```
Might be better to raise a ConanException, so the user knows that the path for profiles (we will also add this functionality in 0.17 release), instead of silently showing there are not profiles

memsharded commented:
I think we can make it into the next 0.17 release, but lets wait for @lasote feedback. Thanks very much for your contribution!

mropert commented:
Thanks!
I would have added a few tests but my systems are Windows and FreeBSD, not able to cleanly run tests at the time :(
I have a few patches to make FreeBSD work also if you're interested, but more work is still required to pass all tests.

memsharded commented:
Hi @mropert ,

No worries, I'll contribute some test to your PR. 

About the support for FreeBSD, of course, try to open issues, split the task if very big in smaller ones, and we will help with  the testing, thanks!

mropert commented:
Sure, will do

memsharded commented:
Hi Mathieu,

I am writing a couple of tests for this functionality and wonder about the proposed naming for ``conan profile show``. You are using a new ``[sections]`` name to output some scopes, not fully sure why. I think that maybe the best would be to just output the ``profile.dumps()``. In that way the user clearly identifies each thing with their model they have in mind for profiles, not a different representation. If it is matter of coloring the profile, I think it can be done without modifying the contents, I can code an alternative. What do you think?

mropert commented:
The [sections] is a leftover from a first trial that I should have renamed.
I actually meant [options] of course.

As for dumps(), at first I reused it but I found it lacked the pretty colours other commands have :)
So I made my own printer...

memsharded commented:
Ok, great. I might try to do colored output with another approach, or just change the name. Thanks!

mropert commented:
I've rebased, fixed the label to [scopes] and corrected the called to .items() in the scope loop by the way.
Travis failed but I'm unsure it's related to my new change.

memsharded commented:
I am very sorry, but the fact is that your PR has collided with recent changes in the Profiles feature. Now profiles can be stored anywhere, as a means to let developers share their profiles. So ``--profile=xxx```, now xxx can be an absolute path, a relative path, or a name from the user client_cache.

The ``read_profile`` functionality has been refactored to: https://github.com/conan-io/conan/blob/develop/conans/client/manager.py#L222

So the functions that you are adding in the client_cache shouldn't be used anymore.

But don't worry :) Your PR is good, you did a good job. I can propose a refactor to your PR tonight to match the new profiles, so you can merge it into yours, and it will get into the next 0.17 release. BTW CI in travis was stalled, not your PR fault, I just re-launched. THanks very much!

memsharded commented:
Hi Mathieu,

Finally we merged both your PR and my PR with the tests and the small changes to output. We thought it was better to have them in the 0.17 release, and we can iterate and modify the behavior later, but the commands are already useful as they are, so it was an interesting feature. Thanks very much again for your contribution!

mropert commented:
No problem, sorry I couldn't reply earlier.
In the future, I think we might want to find a way to only show what the user set in the config and not what Conan implied (the "Dev=true" thing).
That might require to split the Profile into two: the original one read from the config file and the one Conan client uses (enriched with the default settings).
